katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+################################################################################
+# Copyright (c) 2017-2018, National Research Foundation (Square Kilometre Array)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Function to apply TelescopeState calibration to visibilities.
+
+Hold, interpolate and apply cal solutions from a katdal object.
+Code largely pilfered from the katdal_loader in katsdpimager, but using a katdal object
+rather than a filename, so that things are a bit more portable. """
+
+import dask.array as da
+import numpy as np
+
+from .lazy_indexer import LazyTransform
+from .sensordata import _safe_linear_interp
+
+
+class CalibrationReadError(RuntimeError):
+    """An error occurred in loading calibration values from file"""
+    pass
+
+
+class CalibrationValueError(ValueError):
+    """An error occurred in the calculation of calibration coefficients"""
+    pass
+
+
+class SimpleInterpolate1D(object):
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+
+    def __call__(self, x):
+        y_re = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        y_im = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        # Interpolate across nans in time axis.
+        for ch_idx in range(self._y.shape[1]):
+            for pol_idx in range(self._y.shape[2]):
+                for ant_idx in range(self._y.shape[3]):
+                    y_re[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                            self._y[:, ch_idx, pol_idx, ant_idx].real,
+                                                                            x)
+                    y_im[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                            self._y[:, ch_idx, pol_idx, ant_idx].imag,
+                                                                            x)
+        y = y_re + y_im
+        return y
+
+
+class ComplexInterpolate1D(object):
```

Comment:
The names of SimpleInterpolate1D and ComplexInterpolate1D are a bit misleading, because both interpolate complex numbers.